coding-harder/TreeTraversal.js

```diff
@@ -0,0 +1,287 @@
+      else if (key > x.key) x = x.right // traverse right
+      else return x.value // found it!
+    }
+    return x
+  }
+
+  delete(key) {
+    function _delete(node, key) {
+      if (node === null) return
+      if (key < node.key) {
+        node.left = _delete(node.left, key)
+      } else if (key > node.key) {
+        node.right = _delete(node.right, key)
+      } else { // node.key === key
+        if (node.right === null) return node.left
+        if (node.left === null) return node.right
+
+        let t = node
+        node = min(t.right)
+        node.right = removeMin(t.right)
+        node.left = t.left
+      }
+      return node
+    }
+
+    function removeMin(node) {
+      if (node.left === null) return node.right
+      node.left = removeMin(node.left)
+      return node
+    }
+
+    function min(node) {
+      let x = node
+      while (x.left !== null) x = x.left
+      return x
+    }
+
+    this.root = _delete(this.root, key)
+  }
+
+  contains(key) {
+    return this.get(key) !== null
+  }
+
+  isEmpty() {
+    return this.root === null
+  }
+
+  size() {
+    function _size(node) {
+      return node === null ? 0 : 1 + _size(node.left) + _size(node.right)
+    }
+    return _size(this.root)
+  }
+
+  inorder() {
+    let queue = new Queue()
+    function _inorder(node) {
+      if (node === null) return
+      _inorder(node.left)
+      queue.enqueue(node)
+      _inorder(node.right)
+    }
+    _inorder(this.root)
+    return queue
+  }
+
+  preorder() {
+    let queue = new Queue()
+    function _preorder(node) {
+      if (node === null) return
+      queue.enqueue(node)
+      _preorder(node.left)
+      _preorder(node.right)
+    }
+    _preorder(this.root)
+    return queue
+  }
+
+  postorder() {
+    let queue = new Queue()
+    function _postorder(node) {
+      if (node === null) return
+      _postorder(node.left)
+      _postorder(node.right)
+      queue.enqueue(node)
+    }
+    _postorder(this.root)
+    return queue
+  }
+
+  bfs() {
+    if (this.root === null) return new Queue()
+    let queue = new Queue()
+    let nodes = new Queue()
+    queue.enqueue(this.root)
+    while (!queue.isEmpty()) {
+      let node = queue.dequeue()
+      nodes.enqueue(node)
+      if (node.left !== null) queue.enqueue(node.left)
+      if (node.right !== null) queue.enqueue(node.right)
+    }
+    return nodes
+  }
+}
+
+class Node {
+  constructor(key, value) {
+    this.key = key
+    this.value = value
+    this.left = null
+    this.right = null
+  }
+}
+
+class Queue {
+  constructor(...items) {
+    this.data = []
+    for (let item of items) {
+      this.enqueue(item)
+    }
+  }
+
+  enqueue(item) { this.data.unshift(item) }
+  dequeue() { return this.data.pop() }
+  isEmpty() { return this.data.length === 0 }
+
+  [Symbol.iterator]() {
+    let idx = this.data.length - 1
+    return {
+      next: () => idx > -1
+        ? { value: this.data[idx--], done: false }
+        : { done: true }
+    }
+  }
```

Comment:
This is really nice! You might consider making this a generator and using `yield` to reduce some of this boilerplate!